radix-engine/src/engine/id_allocator.rs

```diff
@@ -1,45 +1,142 @@
+use crate::engine::{KernelError, RuntimeError};
 use radix_engine_interface::api::types::{
-    AuthZoneStackId, BucketId, ComponentId, FeeReserveId, KeyValueStoreId, NonFungibleStoreId,
-    PackageId, ProofId, ResourceManagerId, TransactionRuntimeId, ValidatorId, VaultId,
+    AuthZoneStackId, BucketId, ComponentId, FeeReserveId, GlobalAddress, KeyValueStoreId,
+    NonFungibleStoreId, PackageId, ProofId, RENodeId, RENodeType, ResourceManagerId,
+    TransactionRuntimeId, ValidatorId, VaultId,
 };
 use radix_engine_interface::crypto::{hash, Hash};
 use radix_engine_interface::model::*;
-use sbor::rust::ops::Range;
+use sbor::rust::collections::BTreeSet;
+use sbor::rust::string::ToString;
+use sbor::rust::vec;
+use sbor::rust::vec::Vec;
 
 use super::IdAllocationError;
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub enum IdSpace {
-    System,
-    Application,
-}
-
 /// An ID allocator defines how identities are generated.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct IdAllocator {
-    available: Range<u32>,
+    pre_allocated_ids: BTreeSet<RENodeId>,
+    frame_allocated_ids: Vec<BTreeSet<RENodeId>>,
+    next_id: u32,
     transaction_hash: Hash,
 }
 
 impl IdAllocator {
     /// Creates an ID allocator.
-    pub fn new(kind: IdSpace, transaction_hash: Hash) -> Self {
+    pub fn new(transaction_hash: Hash, pre_allocated_ids: BTreeSet<RENodeId>) -> Self {
         Self {
-            available: match kind {
-                IdSpace::System => 0..1024,
-                IdSpace::Application => 1024..u32::MAX,
-            },
+            pre_allocated_ids,
+            frame_allocated_ids: vec![BTreeSet::new()],
+            next_id: 0u32,
             transaction_hash,
         }
     }
 
-    fn next(&mut self) -> Result<u32, IdAllocationError> {
-        if self.available.len() > 0 {
-            let id = self.available.start;
-            self.available.start += 1;
-            Ok(id)
+    pub fn pre_execute_invocation(&mut self) {
+        self.frame_allocated_ids.push(BTreeSet::new());
+    }
+
+    pub fn post_execute_invocation(&mut self) -> Result<(), RuntimeError> {
+        if let Some(ids) = self.frame_allocated_ids.pop() {
+            if !ids.is_empty() {
+                return Err(RuntimeError::KernelError(KernelError::IdAllocationError(
+                    IdAllocationError::AllocatedIDsNotEmpty,
+                )));
+            }
+            Ok(())
+        } else {
+            Err(RuntimeError::UnexpectedError("No frame found.".to_string()))
```

Comment:
Should just panic?

Reply:
I've started replacing our panics with `RuntimeError::UnexpectedError` so that we leave it up to user of the lib (in our case `babylon-node`) to panic or not. Thoughts on that?

Reply:
I'm still leaning towards panic internally if our core assumption breaks, so we can distinguish implementation failures from  user errors. I'm not sure how much leeway node has, if engine can't handle it properly.

Reply:
If core assumptions are broken then a node could theoretically just abort execution and not commit anything to database. But I'm fine with also panicking.

Reply:
Yeah, I understand that viewpoint. 

I think we can gain more from panics by not allowing the network to proceed with possibly wrong implementaiton (say some smart node continues to run by skipping "illegal" transactions) and getting presummably simpler interface.

The other approach would require a rigirious review of our integration points and make sure all places are filtering error code properlys, to pause and/or report in case of expected error.

Reply:
Perhaps it makes sense to propagate up these errors, but (for now) convert them into a panic during finalization? (with the goal of turning that into a runtime configuration option?)

Reply:
Will go with Yulong's approach for now as converting them into a panic during finalization does have weird abstraction issues where client would still need to handle the `RuntimeError::UnexpectedError` case. We can revisit this later :)